cdc/processor/sinkmanager/manager.go

```diff
@@ -345,90 +379,108 @@ func (m *SinkManager) generateSinkTasks() error {
-				var upperBoundTs model.Ts
-				if resolvedTs > barrierTs {
-					upperBoundTs = barrierTs
-				} else {
-					upperBoundTs = resolvedTs
-				}
+			tables = append(tables, tableSink)
+			progs = append(progs, slowestTableProgress)
+		}
 
-				return engine.Position{
-					StartTs:  upperBoundTs - 1,
-					CommitTs: upperBoundTs,
-				}
-			}
-			upperBound := getUpperBound()
-			// Only generate the table sink task if lower bound less or equal the upper bound.
-			checkAdvance := slowestTableProgress.nextLowerBoundPos.Compare(upperBound)
-			if !(checkAdvance == -1 || checkAdvance == 0) || !m.memQuota.tryAcquire(requestMemSize) {
+		i := 0
+	LOOP:
+		for ; i < len(tables); i++ {
+			tableSink := tables[i]
+			slowestTableProgress := progs[i]
+			lowerBound := slowestTableProgress.nextLowerBoundPos
+			upperBound := getUpperBound(tableSink)
+
+			// The table has no available progress.
+			if lowerBound.Compare(upperBound) >= 0 {
 				m.sinkProgressHeap.push(slowestTableProgress)
-				// Next time.
 				continue
 			}
+
+			// No available memory, skip this round directly.
+			if !m.memQuota.tryAcquire(requestMemSize) {
+				break LOOP
+			}
+
 			log.Debug("MemoryQuotaTracing: Acquire memory for table sink task",
 				zap.String("namespace", m.changefeedID.Namespace),
 				zap.String("changefeed", m.changefeedID.ID),
-				zap.Int64("tableID", tableID),
-				zap.Uint64("memory", requestMemSize),
-			)
-			callback := func(lastWrittenPos engine.Position) {
-				p := &progress{
-					tableID:           tableID,
-					nextLowerBoundPos: lastWrittenPos.Next(),
-				}
-				m.sinkProgressHeap.push(p)
-			}
+				zap.Int64("tableID", tableSink.tableID),
+				zap.Uint64("memory", requestMemSize))
 
 			t := &sinkTask{
-				tableID:       tableID,
-				lowerBound:    slowestTableProgress.nextLowerBoundPos,
+				tableID:       tableSink.tableID,
+				lowerBound:    lowerBound,
 				getUpperBound: getUpperBound,
-				tableSink:     tableSink.(*tableSinkWrapper),
-				callback:      callback,
+				tableSink:     tableSink,
+				callback: func(lastWrittenPos engine.Position) {
+					p := &progress{tableID: tableSink.tableID, nextLowerBoundPos: lastWrittenPos.Next()}
+					m.sinkProgressHeap.push(p)
+					select {
+					case m.sinkWorkerAvailable <- struct{}{}:
+					default:
+					}
+				},
 				isCanceled: func() bool {
-					return tableSink.(*tableSinkWrapper).getState() != tablepb.TableStateReplicating
+					return tableSink.getState() != tablepb.TableStateReplicating
 				},
 			}
 			select {
 			case <-m.ctx.Done():
 				return m.ctx.Err()
 			case m.sinkTaskChan <- t:
+				log.Debug("Generate sink task",
+					zap.String("namespace", m.changefeedID.Namespace),
+					zap.String("changefeed", m.changefeedID.ID),
+					zap.Int64("tableID", tableSink.tableID),
+					zap.Any("lowerBound", lowerBound),
+					zap.Any("currentUpperBound", upperBound))
+			default:
+				m.memQuota.refund(requestMemSize)
+				break LOOP
```

Comment:
`m.sinkTaskChan` is an unbuffered channel, if sink worker is running in `handleTask`, this sink task will be ignored, is it as expected or something will be lost?

Reply:
It's expected. The task will be return into `progressHeap`.